hpx/lcos/detail/promise_base.hpp

```diff
@@ -0,0 +1,222 @@
+                this->shared_state_.reset();
+            }
+
+            promise_base& operator=(promise_base&& other) HPX_NOEXCEPT
+            {
+                base_type::operator=(std::move(other));
+                id_retrieved_      = other.id_retrieved_;
+                id_                = std::move(other.id_);
+                addr_              = std::move(other.addr_);
+
+                other.id_   = naming::invalid_id;
+                other.addr_ = naming::address();
+                return *this;
+            }
+
+            naming::id_type get_id(error_code& ec = throws) const
+            {
+                if (this->shared_state_ == 0)
+                {
+                    HPX_THROWS_IF(ec, no_state,
+                        "detail::promise_base<Result, RemoteResult>::get_id",
+                        "this promise has no valid shared state");
+                    return naming::invalid_id;
+                }
+                if (!addr_ || !id_)
+                {
+                    HPX_THROWS_IF(ec, no_state,
+                        "detail::promise_base<Result, RemoteResult>::get_id",
+                        "this promise has no valid LCO");
+                    return naming::invalid_id;
+                }
+                if (!this->future_retrieved_)
+                {
+                    HPX_THROW_EXCEPTION(invalid_status,
+                        "promise<Result>::get_id",
+                        "future has not been retrieved from this promise yet");
+                    return naming::invalid_id;
+                }
+
+                id_retrieved_ = true;
+
+                return id_;
+            }
+
+            naming::id_type get_unmanaged_gid(error_code& ec = throws) const
+            {
+                return get_id(ec);
+            }
+
+#if defined(HPX_HAVE_COMPONENT_GET_GID_COMPATIBILITY)
+            naming::id_type get_gid(error_code& ec = throws) const
+            {
+                return get_id(ec);
+            }
+#endif
+
+            naming::address resolve(error_code& ec = throws) const
+            {
+                if (!addr_ || !id_)
+                {
+                    HPX_THROWS_IF(ec, no_state,
+                        "detail::promise_base<Result, RemoteResult>::get_id",
+                        "this promise has no valid LCO");
+                    return naming::address();
+                }
+                return addr_;
+            }
+
+        protected:
+            void check_abandon_shared_state(const char* fun)
+            {
+                if (this->shared_state_ != 0 && this->future_retrieved_ &&
+                    !(this->has_result_ || id_retrieved_))
+                {
+                    this->shared_state_->set_error(broken_promise, fun,
+                        "abandoning not ready shared state");
+                    return;
+                }
+            }
+            mutable bool id_retrieved_;
+
+            naming::id_type id_;
```

Comment:
It would. But we need the id_type currently eventually, so why not create it right away?